### PR TITLE
improve performance of getting hierarchy

### PR DESCRIPTION
**Pull request by tpluscode**
I realise that this became a big PR but most in dev-deps and test code

Bottom line, I first build a SHACL Shape to represent the hierarchy and then generate a query using [@hydrofoil/shape-to-query](https://www.npmjs.com/package/@hydrofoil/shape-to-query)

Will add more comments inline

### CODE REVIEW

File: lib/hierarchyShape.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    term: hierarchy.term,
+  })
+
+  const rootShape = clone.blankNode()
+
+  rootShape
+    .addOut(rdf.type, sh.NodeShape)
+    .addOut(sh.targetNode, hierarchy.out(meta.hierarchyRoot))
+    .addOut(sh.rule, rule => {
+      rule.addOut(rdf.type, s2q.SPORule)
```
Comment (tpluscode) [inline lib/hierarchyShape.ts:19]: To mimic `DESCRIBE`, I add this [SPO Rule](https://shape-to-query.hypermedia.app/docs/#extension-spo-rule) which generates open-ended `?s ?p ?o` patterns to every hierarchy level

File: lib/hierarchyShape.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+          })
+
+        const targetClass = currentLevel.out(sh.targetClass).term
+        if (targetClass) {
+          constraint.addOut(sh.class, targetClass)
+        }
+
+        if (isNamedNode(path)) {
+          currentLevelShape.out(sh.rule)
+            .addOut(s2q.predicateFilter, propertyFilter(path))
```
Comment (tpluscode) [inline lib/hierarchyShape.ts:51]: This is to exclude the hierarchy property from `?s ?p ?o` which is necessary when the hierarchy level is constrained. See https://github.com/zazuko/cube-hierarchy-query/issues/17#issuecomment-1562541933

File: test/lib/__snapshots__/hierarchyShape.test.ts.snap [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+					) ;
+				] ;
+			] ;
+			sh:property [
+				sh:path schema:containsPlace ;
+			] ;
+		] ;
+	] .
+
+"
```
Comment (tpluscode) [inline test/lib/__snapshots__/hierarchyShape.test.ts.snap:80]: This is what the generated SHACL Shape looks like. You can try it in https://shape-to-query.hypermedia.app/

**Review by ludovicm67 — APPROVED** [review]
I'm not sure if I'm the best one for reviewing this, since I'm not aware of all that libraries on detail.

The overall looks good, and the performance measurements are great to have ; this will definitively help on improving things here.

Since tests are passing and that it looks good, I will approve this.

File: examples/hierarchy.ts [hunk 4]
```diff
@@ -33,9 +42,15 @@ const main = async () => {
     throw new Error(`Hierarchy not found ${args.dimensionIri}`)
   }
 
+  performance.mark('begin getHierarchy')
   const results = await getHierarchy(hierarchy).execute(client, $rdf)
+  performance.mark('end getHierarchy')
+  performance.measure('getHierarchy', 'begin getHierarchy', 'end getHierarchy')
 
   results.forEach(print(0))
+  performance.measure('print', 'begin print', 'end print')
```
Comment (Rdataflow) [inline examples/hierarchy.ts]: @tpluscode detail: this line leads to ...
```
DOMException [SyntaxError]: The "end print" performance mark has not been set
    at new DOMException (node:internal/per_context/domexception:53:5)
    at __node_internal_ (node:internal/util:663:10)
    at getMark (node:internal/perf/usertiming:68:11)
    at calculateStartDuration (node:internal/perf/usertiming:183:11)
    at measure (node:internal/perf/usertiming:211:7)
    at Performance.measure (node:internal/perf/performance:135:12)
    at main (file:///C:/temp/cube-hierarchy-query/examples/hierarchy.ts:44:17)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at runNextTicks (node:internal/process/task_queues:64:3)
    at listOnTimeout (node:internal/timers:540:9)
```

Comment (Rdataflow) [reply examples/hierarchy.ts]: ... apart from above detail - the performance of the resulting query now looks really performant! THX
I'm already curious to see it released to the wild :-)

File: .changeset/quick-toes-lie.md [hunk 5]
```diff
@@ -0,0 +1,5 @@
+---
+"@zazuko/cube-hierarchy-query": major
+---
+
+The `getHierarchy` exported by the main module now generates a `CONSTRUCT` query which should yield a great improvement in performance in some cases (closes #17)
```
Comment (giacomociti) [inline .changeset/quick-toes-lie.md:5]: in the readme, a 'DESCRIBE' query it is still mentioned

**Review by ludovicm67 — APPROVED** [review]

**Comment by Rdataflow** [comment]
@tpluscode well done 🎉

**Pull request merged**